src/librustc_trans/trans/mir/operand.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright 2012-2014 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use llvm::ValueRef;
+use rustc::middle::ty::Ty;
+use rustc_mir::repr as mir;
+use trans::base;
+use trans::build;
+use trans::common::Block;
+use trans::datum;
+
+use super::MirContext;
+
+pub struct OperandRef<'tcx> {
+    // This will be "indirect" if `appropriate_rvalue_mode` returns
+    // ByRef, and otherwise ByValue.
+    pub llval: ValueRef,
+
+    // The type of value being returned.
+    pub ty: Ty<'tcx>
+}
+
+impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
+    pub fn trans_operand(&self,
+                         bcx: Block<'bcx, 'tcx>,
+                         operand: &mir::Operand<'tcx>)
+                         -> OperandRef<'tcx>
+    {
+        match *operand {
+            mir::Operand::Consume(ref lvalue) => {
+                let tr_lvalue = self.trans_lvalue(bcx, lvalue);
+                let ty = tr_lvalue.ty.to_ty(bcx.tcx());
+                let llval = match datum::appropriate_rvalue_mode(bcx.ccx(), ty) {
+                    datum::ByValue => build::Load(bcx, tr_lvalue.llval),
+                    datum::ByRef => tr_lvalue.llval,
+                };
+                OperandRef {
+                    llval: llval,
+                    ty: ty
+                }
+            }
+
+            mir::Operand::Constant(ref constant) => {
+                let llval = self.trans_constant(bcx, constant);
+                let ty = bcx.monomorphize(&constant.ty);
+                OperandRef {
+                    llval: llval,
+                    ty: ty,
+                }
+            }
+        }
+    }
+
+    pub fn trans_operand_into(&self,
+                              bcx: Block<'bcx, 'tcx>,
+                              lldest: ValueRef,
+                              operand: &mir::Operand<'tcx>)
+    {
+        match *operand {
+            mir::Operand::Consume(ref lvalue) => {
+                let tr_lvalue = self.trans_lvalue(bcx, lvalue);
+                let lvalue_ty = tr_lvalue.ty.to_ty(bcx.tcx());
+                base::memcpy_ty(bcx, lldest, tr_lvalue.llval, lvalue_ty);
```

Comment:
if `mir::Operand::Consume` stands for owned value, shouldn’t this be using memmove?